docs/openstack/node-selector.md

```diff
@@ -52,7 +52,7 @@ metadata:
   name: openstack
 spec:
   secret: osp-secret
-  storageClass: local-storage
+  storageClass: crc-csi-hostpath-provisioner
```

Comment:
shouldn't this use `$STORAGE_CLASS_NAME` as the other files?

Reply:
this document looks detached from the ohter places and there is no env variables for it, so I am not sure if it should, or this is good as is. @Akrog WDYT?

Reply:
The CR should always have `storageClass: local-storage`.

Reply:
Today, `install_yamls` provides a `local-storage` class which we can use for testing. It uses `make crc_storage` for now. 

In the future `ci-framework` will provide `local-storage` and it will be implemented via LVMS. I hope everyone will be on ci-framework by then, but I expect I'll update install_yamls too so that it has something like `make lvms_storage`.

Reply:
so when customer deploy this in production i would not expect them to use a storage class called "local-storage" in general. themy may but i would expect them to use a storage class that is preconfigured in there openshift which may have another name.

we discussed this a bit on slack and I'm not convinced we whoudl move our default testing to lvms.

we use crc-csi-hostpath-provisioner in ci for non adoption jobs because make crc storage was unreliable in the past when using the nested virt crc.

that may have resolved it self with the move to split out crc.

to me deviating form crc-csi-hostpath-provisioner in the zuul based jobs is an unnecessary risk and maintenance burden so i would reserve testing with lvms for baremetal jobs that do not use crc.
i.e. jobs that use SNO or similar.

I'm not convinced that we should have make lvms_storage target in install yamls. we could but presumably that would need to either
modify the CRC VM with an additional disk to use for the lvm topology operator and install and configure that operator
or it would have to create a file in the CRC VM and mount that as a loopback device which is then used by the lvm topology operator.

the loop back device concerns me due to the io overhead with that configuration. we do use that for cinder lvm with devstack but my experience with the nova gate is we do have io issues in the lvm based jobs when they run on a slow node. i would be concerned with recreating that problem in our zuul ci with CRC.

our operator ci is sufficiently different that perhaps we wont see the lvm issues we see with loop back devices there however its a concern. I'm not convinced that modifying the ci-framework to add cinder volumes to the zuul vms is a good approach either nor do i think modifying the CRC bundle image to add an additional portion that can be used for the lvm topology operator is a good approach. 
both add technical debt that i think we we would be better to avoid.

unless the is a very strong reason not continue to use crc-csi-hostpath-provisioner as our storage class for all crc based zull jobs i would be strongly in favor of adopting that for the adoption jobs too.

Reply:
addressed @fultonj 's @cescgina 's comments

Reply:
> so when customer deploy this in production i would not expect them to use a storage class called "local-storage" in general. themy may but i would expect them to use a storage class that is preconfigured in there openshift which may have another name.
> 
> we discussed this a bit on slack and I'm not convinced we whoudl move our default testing to lvms.

We don't have to resolve this debate in order to merge this patch. I'm OK with merging it as @bogdando has right now and we can talk more about LVMS vs other backends. I think it depends on the job.

> we use crc-csi-hostpath-provisioner in ci for non adoption jobs because make crc storage was unreliable in the past when using the nested virt crc.
> 
> that may have resolved it self with the move to split out crc.
> 
> to me deviating form crc-csi-hostpath-provisioner in the zuul based jobs is an unnecessary risk and maintenance burden so i would reserve testing with lvms for baremetal jobs that do not use crc. i.e. jobs that use SNO or similar.

I wouldn't limit it only to baremetal. There are going to be jobs deployed by ci-framework which use devscripts-deployed OCP on VMs. I think LVMS would be good for them to test with.

For the [HCI VA](https://github.com/openstack-k8s-operators/architecture/tree/main/examples/va/hci) I would like to move the control-planes PVCs to LVMS and I'd like that also for jobs which will test it, specifically ci-framework jobs. I'm less concerned with crc jobs for that.

> I'm not convinced that we should have make lvms_storage target in install yamls. we could but presumably that would need to either modify the CRC VM with an additional disk to use for the lvm topology operator and install and configure that operator or it would have to create a file in the CRC VM and mount that as a loopback device which is then used by the lvm topology operator.
> 
> the loop back device concerns me due to the io overhead with that configuration. we do use that for cinder lvm with devstack but my experience with the nova gate is we do have io issues in the lvm based jobs when they run on a slow node. i would be concerned with recreating that problem in our zuul ci with CRC.
> 
> our operator ci is sufficiently different that perhaps we wont see the lvm issues we see with loop back devices there however its a concern. I'm not convinced that modifying the ci-framework to add cinder volumes to the zuul vms is a good approach either nor do i think modifying the CRC bundle image to add an additional portion that can be used for the lvm topology operator is a good approach. both add technical debt that i think we we would be better to avoid.
> 
> unless the is a very strong reason not continue to use crc-csi-hostpath-provisioner as our storage class for all crc based zull jobs i would be strongly in favor of adopting that for the adoption jobs too.

Reply:
> so when customer deploy this in production i would not expect them to use a storage class called "local-storage" in general. themy may but i would expect them to use a storage class that is preconfigured in there openshift which may have another name.

I think we should have a storage class recommendation for customers to use in production which uses the local storage of the OCP worker nodes and does not require them to purchase an additional storage solution. To require such a purchase would be hardware scope creep because we have supported the local disks of our controller nodes in previous OpenStacks for control plane services. I think we should use LVMS to implement this. I also want to ensure that any limitation from using this local storage (e.g. no support for RWX, but other issues could exist)  does not affect our control plane. I.e. we need to test what we recommend. I would call this storage class "local-storage" but it could be called anything. In our HCI VA its name is kustomizable:

https://github.com/openstack-k8s-operators/architecture/blob/main/examples/va/hci/values.yaml#L201
https://github.com/openstack-k8s-operators/architecture/blob/main/examples/va/hci/values-ci-framework.yaml#L216

I can imagine useful crc-based jobs which don't use LVMS too so we don't need to use it in 100% of the jobs, but if the control plane does not work with LVMS I'd consider it a blocker.

Reply:
> if it did not work with lvms it would be a regression

agreed.

> perhaps we shoudl expect operator lint with a check to catch any direct request for a PVC or type RWX and omit a warning or error? that shoudl prevent us from regressing?

That's an interesting idea. Do you know of any existing patterns which have that implemented?

Reply:
so @gibizer wrote the tool but i have not looked tat the details

https://github.com/gibizer/operator-lint

here is an example rule for crds
https://github.com/gibizer/operator-lint/tree/main/linters/crd/C003

you can see its using the golang ast tooling to inspect the code

https://github.com/gibizer/operator-lint/blob/main/linters/crd/C003/C003.go#L36-L50

i haven't played with that interface before but i suspect it would not be that difficult to check for the access mode

https://github.com/openstack-k8s-operators/glance-operator/blob/main/pkg/glance/pvc.go#L42-L45

we would need to parse that and validate the access mode

gibi will be back on Monday so perhaps we can discuss it with them next week and see how feasible that is